tester: bas: Add support for testing Battery Service

MicImbriani opened the pull request:
Currently only support mandatory Battery Level and optional Level Status Characteristic.

Signed-off-by: Michele Imbriani <mbim@demant.com>

Thalley commented on tests/bluetooth/tester/src/btp/btp.h:
```diff
@@ -38,67 +38,69 @@
-#define BTP_MTU 1024
+#define BTP_MTU           1024
 #define BTP_DATA_MAX_SIZE (BTP_MTU - sizeof(struct btp_hdr))
 
-#define BTP_INDEX_NONE		0xff
-#define BTP_INDEX		0x00
+#define BTP_INDEX_NONE 0xff
+#define BTP_INDEX      0x00
```
Suggest to undo these changes. Have you set your formatter to format all lines, and not just changed lines?

Thalley replied:
Applies to several other lines in this file

MicImbriani replied:
I didn't set it myself, so perhaps that's the default formatting option?

Anyways, will do, thanks for the suggestion :)

Thalley replied:
> I didn't set it myself, so perhaps that's the default formatting option?

You may have misconfigured your IDE to format lines not changes by you - Most IDE can be configured to only format using e.g. clang-format on lines you've manually modified.

Thalley commented on tests/bluetooth/tester/src/btp/btp_bas.h:
```diff
@@ -0,0 +1,102 @@
+#include <stdint.h>
+#include <stddef.h>
+
+/* BTP BAS Service Opcodes */
+#define BTP_BAS_SET_BATTERY_LEVEL             0x02
+#define BTP_BAS_BLS_SET_BATTERY_PRESENT       0x03
```
Why do they start from `0x02`?

MicImbriani replied:
Because in AutoPTS, 0x01 is reserved for XXX_READ_SUPPORTED_COMMANDS.

Thalley replied:
Aha. Shouldn't we define and implement a `BTP_BASS_READ_SUPPORTED_COMMANDS` then?

Thalley replied:
@sjanc What's your input here? :)

sjanc replied:
it should be implemented,  although currently autopts is not making use of it we have a long term goal to start using that to provide more generic WID handlers based on what is supported by IUT 

and it is just a few lines of code, you can just copy paste implementation from other services and update bitmask for BAS commands

Thalley commented on tests/bluetooth/tester/src/btp/btp_bas.h:
```diff
@@ -0,0 +1,102 @@
+	uint8_t reason;
+} __packed;
+
+#if defined(CONFIG_BT_BAS_BLS_IDENTIFIER_PRESENT)
+struct btp_bas_bls_set_identifier_cmd {
+	bt_addr_le_t address;
+	uint16_t identifier;
+} __packed;
+#endif /* CONFIG_BT_BAS_BLS_IDENTIFIER_PRESENT */
+
+#if defined(CONFIG_BT_BAS_BLS_ADDITIONAL_STATUS_PRESENT)
+struct btp_bas_bls_set_service_required_cmd {
+	bt_addr_le_t address;
+	uint8_t service_required;
+} __packed;
+
+struct btp_bas_bls_set_battery_fault_cmd {
+	bt_addr_le_t address;
+	uint8_t battery_fault;
+} __packed;
+#endif /* CONFIG_BT_BAS_BLS_ADDITIONAL_STATUS_PRESENT */
```
```suggestion
struct btp_bas_bls_set_identifier_cmd {
	bt_addr_le_t address;
	uint16_t identifier;
} __packed;

struct btp_bas_bls_set_service_required_cmd {
	bt_addr_le_t address;
	uint8_t service_required;
} __packed;

struct btp_bas_bls_set_battery_fault_cmd {
	bt_addr_le_t address;
	uint8_t battery_fault;
} __packed;
```

No value in guarding definitions :)

Thalley commented on tests/bluetooth/tester/src/btp/btp_bas.h:
```diff
@@ -0,0 +1,102 @@
+/**
+ * @brief Initialize the Battery Service (BAS) tester.
+ *
+ * @return uint8_t Status of the initialization. Returns `BTP_STATUS_SUCCESS`
+ *         if the initialization is successful. Other values indicate failure.
+ */
+uint8_t tester_init_bas(void);
+
+/**
+ * @brief Unregister the Battery Service (BAS) tester.
+ *
+ * @return uint8_t Status of the unregistration. Returns `BTP_STATUS_SUCCESS`
+ *         if the unregistration is successful. Other values indicate failure.
```
```suggestion
 * @return Status of the initialization. Returns `BTP_STATUS_SUCCESS`
 *         if the initialization is successful. Other values indicate failure.
 */
uint8_t tester_init_bas(void);

/**
 * @brief Unregister the Battery Service (BAS) tester.
 *
 * @return Status of the unregistration. Returns `BTP_STATUS_SUCCESS`
 *         if the unregistration is successful. Other values indicate failure.
```

Thalley commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,237 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
```
```suggestion

```

Thalley commented on tests/bluetooth/tester/src/btp/btp_bas.h:
```diff
@@ -0,0 +1,102 @@
+#define BTP_BAS_BLS_SET_BATTERY_FAULT         0x0C
+
+/* Command structures */
+struct btp_bas_set_battery_level_cmd {
+	bt_addr_le_t address;
+	uint8_t level;
+} __packed;
```
Why is `address` a parameter in these? 

It's for a GATT server device, right? Is there any use case for setting a specific value for specific remote devices?

sjanc replied:
yeah, this feels like it should be global value

BTW I'll do BTP review in next few days (currently I'm preempted with other non-zephyr work)

Thalley commented on tests/bluetooth/tester/overlay-le-audio.conf:
```diff
@@ -149,3 +149,10 @@ CONFIG_BT_TBS_SUPPORTED_FEATURES=3
 
 # TMAP
 CONFIG_BT_TMAP=y
+
+#BAS
+CONFIG_BT_BAS=y
+CONFIG_BT_BAS_BLS=y
+CONFIG_BT_BAS_BLS_IDENTIFIER_PRESENT=y
+CONFIG_BT_BAS_BLS_BATTERY_LEVEL_PRESENT=y
+CONFIG_BT_BAS_BLS_ADDITIONAL_STATUS_PRESENT=y
```
Doesn't this belong in the `prj.conf` file instead of the LE Audio overlay? BAS is not LE Audio related

MicImbriani commented:
I have made a mistake while rebasing, causing this PR to temporarily be in an incorrect state. I am fixing it right now

Thalley commented:
@MicImbriani Looks like your rebase went wrong :)

fgrandel commented:
Removed all but the recommended reviewers.

sjanc commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,266 @@
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_WIRELESS_POWER_SOURCE);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_STATE);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_LEVEL);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_TYPE);
```
this is already octet 1

sjanc commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,266 @@
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_SERVICE_REQUIRED);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_FAULT);
+
+	*rsp_len = sizeof(*rp) + 1;
```
should be +2

sjanc commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,266 @@
+{
+	const struct btp_bas_bls_set_battery_present_cmd *cp =
+		(const struct btp_bas_bls_set_battery_present_cmd *)cmd;
+	enum bt_bas_bls_battery_present present = cp->present;
```
should be validated before pass to function (especially if function is void and doesn't return error)

(same for other handlers)

sjanc commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,266 @@
+{
+	const struct btp_bas_bls_set_identifier_cmd *cp =
+		(const struct btp_bas_bls_set_identifier_cmd *)cmd;
+	uint16_t identifier = cp->identifier;
```
should be converted from LE to host order

MicImbriani replied:
Thanks for all the feedback!
Could you pls point me to an example of another file that has the correct Host order?

sjanc reviewed (changes requested):
HI,

some input validation is missing (proper validation and BTP error makes it easier to figure out in case of failing tests), some endianness conversion is also missing

It would be good to get allowed values (in case of enum or bit like values) as defines in btp_bas.h for each command,  this make it clear on how BTP is converted to Zephyr API (even if those are currently mapped 1:1)

MicImbriani committed:
tester: bas: Add support for testing Battery Service

Currently only support mandatory Battery Level and optional Level Status
Characteristic.

Signed-off-by: Michele Imbriani <mbim@demant.com>

Thalley commented:
@MicImbriani looks a rebase gone wrong :)

Thalley commented on tests/bluetooth/tester/src/btp/btp.h:
```diff
@@ -75,8 +76,9 @@
 #define BTP_SERVICE_ID_TBS	27
 #define BTP_SERVICE_ID_TMAP	28
 #define BTP_SERVICE_ID_OTS	29
+#define BTP_SERVICE_ID_BAS	31
```
30 is PBP, right?

Just be aware that we are modifying this to be hex in https://github.com/zephyrproject-rtos/zephyr/pull/80740

MicImbriani replied:
Correct. 

I see, I will wait for that PR to be merged.

Thalley replied:
I would be OK to also add the PBP service value here, so that it's more obvious why 31 is used for BAS.

Thalley commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,308 @@
+{
+	struct btp_bas_read_supported_commands_rp *rp = rsp;
+
+	/* octet 0 */
+	tester_set_bit(rp->data, BTP_BAS_READ_SUPPORTED_COMMANDS);
+	tester_set_bit(rp->data, BTP_BAS_SET_BATTERY_LEVEL);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_PRESENT);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_WIRED_POWER_SOURCE);
+
+	/* octet 1 */
```
Why are we only using 4 bits in the first octet?

Doesn't `tester_set_bit(rp->data, BTP_BAS_BLS_SET_WIRELESS_POWER_SOURCE);` set it at BIT(5), which is octet 0 and not octet 1?

MicImbriani replied:
This was a suggestion from @sjanc . 
Perhaps I have misunderstood it

Thalley replied:
Since `tester_set_bit` basically just does `BIT()`, any values below 0x08 ends up in octet 0, values from 0x08->0x0F (inclusive) ends up in octet 1

Thalley replied:
So it's just the comment that should be moved so that it matches the values

Thalley commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,308 @@
+		(const struct btp_bas_bls_set_battery_present_cmd *)cmd;
+	enum bt_bas_bls_battery_present present = cp->present;
+
+	if (present != 0 && present != 1) {
+		return BTP_STATUS_FAILED;
+	}
```
Suggest to use the defined values from `bas.h`:
```suggestion
	if (present != BT_BAS_BLS_BATTERY_NOT_PRESENT && present != BT_BAS_BLS_BATTERY_PRESENT) {
		return BTP_STATUS_FAILED;
	}
```

This comment applies to all functions in this PR where applicable

Thalley commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,308 @@
+		(const struct btp_bas_bls_set_wireless_power_source_cmd *)cmd;
+	enum bt_bas_bls_wireless_power_source source = cp->source;
+
+	if (source < BT_BAS_BLS_WIRELESS_POWER_NOT_CONNECTED || source > BT_BAS_BLS_WIRELESS_POWER_UNKNOWN) {
```
This is fine, but you if you want to, you can also consider using the `IN_RANGE` macro:
```suggestion
	if (!IN_RANGE(source, BT_BAS_BLS_WIRELESS_POWER_NOT_CONNECTED BT_BAS_BLS_WIRELESS_POWER_UNKNOWN) {
```

Thalley commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,308 @@
+{
+	const struct btp_bas_bls_set_wireless_power_source_cmd *cp =
+		(const struct btp_bas_bls_set_wireless_power_source_cmd *)cmd;
+	enum bt_bas_bls_wireless_power_source source = cp->source;
```
I'm not 100% sure, but I think that @sjanc previously have disliked casting the BTP values to the enums like this, but not sure.

sjanc replied:
ideally I'd prefer BTP defines and explicit assing, but if enum values are explicitly assigned and match BTP I'm fine with this, but is must be commented and source shall be validated to not be outside of enum values before assigning

MicImbriani replied:
I moved the source verification before the assignment (and in all other applicable places in the file).

Regarding the commenting, I can see why it could be useful, but wouldn't it be redundant to have a comment in these functions when the actual btp_xx_cmd functions are just one search away?

Thalley replied:
I guess the challenge is that the values in BTP are not really defined. 

So for example in `btp_bas_bls_set_wired_power_source` you do `enum bt_bas_bls_wired_power_source source = cp->source;`, but that assumes the the BTP value range and the Zephyr enum value range are identical. 

Zephyr (currently) follows the values defined by the BAS spec, but that's not something that is set in stone, and the values make change at any time (although unlikely). 

Ideally, and this is not specific for this PR or BAS, BTP would either define the value ranges itself (likely leaning on the value specified in the specs), and Zephyr would have BTP -> Zephyr conversion functions, similar to how we convert from and to HCI values from the Zephyr values like 
```c
uint8_t bt_get_phy(uint8_t hci_phy)
{
	switch (hci_phy) {
	case BT_HCI_LE_PHY_1M:
		return BT_GAP_LE_PHY_1M;
	case BT_HCI_LE_PHY_2M:
		return BT_GAP_LE_PHY_2M;
	case BT_HCI_LE_PHY_CODED:
		return BT_GAP_LE_PHY_CODED;
	default:
		return 0;
	}
}

int bt_get_df_cte_type(uint8_t hci_cte_type)
{
	switch (hci_cte_type) {
	case BT_HCI_LE_AOA_CTE:
		return BT_DF_CTE_TYPE_AOA;
	case BT_HCI_LE_AOD_CTE_1US:
		return BT_DF_CTE_TYPE_AOD_1US;
	case BT_HCI_LE_AOD_CTE_2US:
		return BT_DF_CTE_TYPE_AOD_2US;
	case BT_HCI_LE_NO_CTE:
		return BT_DF_CTE_TYPE_NONE;
	default:
		return BT_DF_CTE_TYPE_NONE;
	}
}
```

But that is way outside the scope of this PR

MicImbriani replied:
Thanks for the great explanation. I added the BTP values range as a comment, and I will also update the doc file on the BTP (autopts) side to reflect this.

Thalley reviewed (dismissed):
I'm OK with this

Thalley commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,430 @@
+	struct btp_bas_read_supported_commands_rp *rp = rsp;
+
+	/* octet 0 */
+	tester_set_bit(rp->data, BTP_BAS_READ_SUPPORTED_COMMANDS);
+	tester_set_bit(rp->data, BTP_BAS_SET_BATTERY_LEVEL);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_PRESENT);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_WIRELESS_POWER_SOURCE);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_STATE);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_LEVEL);
+
+	/* octet 1 */
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_IDENTIFIER);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_SERVICE_REQUIRED);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_FAULT);
```
```suggestion
	tester_set_bit(rp->data, BTP_BAS_READ_SUPPORTED_COMMANDS);
	tester_set_bit(rp->data, BTP_BAS_SET_BATTERY_LEVEL);
	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_PRESENT);
	tester_set_bit(rp->data, BTP_BAS_BLS_SET_WIRED_POWER_SOURCE);
	tester_set_bit(rp->data, BTP_BAS_BLS_SET_WIRELESS_POWER_SOURCE);
	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_STATE);
	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_LEVEL);

	/* octet 1 */
	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_TYPE);
	tester_set_bit(rp->data, BTP_BAS_BLS_SET_CHARGING_FAULT_REASON);
	tester_set_bit(rp->data, BTP_BAS_BLS_SET_IDENTIFIER);
	tester_set_bit(rp->data, BTP_BAS_BLS_SET_SERVICE_REQUIRED);
	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_FAULT);
```

Isn't this missing these?

MicImbriani commented:
@sjanc please re-review.

sjanc commented on tests/bluetooth/tester/src/btp/btp_bas.h:
```diff
@@ -0,0 +1,90 @@
+struct btp_bas_bls_set_identifier_cmd {
+	uint16_t identifier;
+} __packed;
+struct btp_bas_bls_set_service_required_cmd {
```
nitpick (empty lines)

sjanc commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,431 @@
+static const struct btp_handler handlers[] = {
+	{
+		.opcode = BTP_BAS_READ_SUPPORTED_COMMANDS,
+		.index = BTP_INDEX,
```
this should probably be BTP_INDEX_NONE

sjanc commented on tests/bluetooth/tester/src/btp_bas.c:
```diff
@@ -0,0 +1,431 @@
+	/* octet 0 */
+	tester_set_bit(rp->data, BTP_BAS_READ_SUPPORTED_COMMANDS);
+	tester_set_bit(rp->data, BTP_BAS_SET_BATTERY_LEVEL);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_PRESENT);
```
some of those handlers are conditional (under ifdefs) so this should also reflect that  (autopts doesn't use it yet, but eventually we want to start relying on supported commands)

sjanc reviewed (commented):
looks ok,  few minor nitpicks
